src/main/java/uk/gov/hmcts/reform/cwrdapi/service/ExcelAdaptorServiceImpl.java

```diff
@@ -0,0 +1,177 @@
+
+        collectHeaderList(headers, sheet);
+        //scan parent and domain object fields by reflection and make maps
+        List<Triple<String,Field, List<Field>>> customObjectFieldsMapping =
+                createBeanFieldMaps(classType, parentFieldMap);
+        Iterator<Row> rowIterator = sheet.rowIterator();
+        rowIterator.next();//skip header
+        while (rowIterator.hasNext()) {
+            Row row = rowIterator.next();
+            Object bean = getInstanceOf(classType.getName());//create parent object
+            for (int i = 0; i < headers.size(); i++) { //set all parent fields
+                setParentFields(getCellValue(row.getCell(i)), bean, headers.get(i), parentFieldMap,
+                        childHeaderToCellMap);
+            }
+            populateChildDomainObjects(bean, customObjectFieldsMapping, childHeaderToCellMap);
+            objectList.add(bean);
+        }
+        return objectList;
+    }
+
+    private void populateChildDomainObjects(
+            Object parentBean, List<Triple<String,Field, List<Field>>> customObjectFields,
+            Map<String, Object> childHeaderValues) {
+        customObjectFields.forEach((customObjectTriple) -> {
+            Field parentField = customObjectTriple.getMiddle();
+            List domainObjectList = new ArrayList();
+            int objectCount = findAnnotation(parentField, MappingField.class).objectCount();//take count from parent
+            for (int i = 0; i < objectCount; i++) {
+                Object childDomainObject = getInstanceOf(customObjectTriple.getLeft());//instantiate child domain object
+                for (Field childField: customObjectTriple.getRight()) {
+                    MappingField mappingField = findAnnotation(childField, MappingField.class);
+                    if (nonNull(mappingField)) {
+                        String domainObjectColumnName = mappingField.columnName().split(DELIMITER_COMMA)[i].trim();
+                        setFieldValue(childField, childDomainObject, childHeaderValues.get(domainObjectColumnName));
+                        setIsPrimaryField(childDomainObject, mappingField, domainObjectColumnName);
+                    }
+                }
+                domainObjectList.add(childDomainObject);//add populated child domain object into list
+            }
+            setFieldValue(parentField, parentBean, domainObjectList);//finally set list to parent field
+        });
+    }
+
+    //called once per file only
+    private List<Triple<String,Field, List<Field>>> createBeanFieldMaps(Class objectClass,
+                                                                  Map<String, Field> headerToCellValueMap) {
+        List<Triple<String,Field, List<Field>>> customObjects = new ArrayList<>();
+        for (Field field: objectClass.getDeclaredFields()) {
+            MappingField mappingField = findAnnotation(field, MappingField.class);
+            if (negate(mappingField.columnName().isEmpty())) {
```

Comment:
mappingField may also nullable here so please include that as well in the condition check?

Reply:
Done